Source/WebCore/html/ImageDataArray.h

```diff
@@ -28,34 +28,147 @@
+    }
+
+    static std::optional<ImageDataArray> tryCreateZeroFilled(size_t length, ImageDataStorageFormat storageFormat)
+    {
+        return tryCreate(length, storageFormat, true);
+    }
+
+    static constexpr bool isSupported(JSC::TypedArrayType typedArrayType)
+    {
+        switch (typedArrayType) {
+        case JSC::TypeUint8Clamped:
+        case JSC::TypeFloat16:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    static constexpr bool isSupported(const JSC::ArrayBufferView& arrayBufferView)
+    {
+        return isSupported(arrayBufferView.getType());
+    }
+
+    static std::optional<ImageDataArray> tryCreate(Ref<JSC::ArrayBufferView>&& arrayBufferView)
+    {
+        if (isSupported(arrayBufferView))
+            return { ImageDataArray(WTFMove(arrayBufferView)) };
+        return { };
+    }
+
     ImageDataArray(Ref<JSC::Uint8ClampedArray>&& data)
-        : m_bufferView(WTFMove(data))
+        : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
     { }
 
+    ImageDataArray(Ref<JSC::Float16Array>&& data)
+        : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+    { }
+
+    ImageDataStorageFormat storageFormat() const
+    {
+        switch (m_bufferView->getType()) {
+        case JSC::TypeUint8Clamped: return ImageDataStorageFormat::Uint8;
+        case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+        default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+        }
+    }
+
+    template <typename F>
+    static auto visit(JSC::ArrayBufferView& bufferView, F&& f)
+    {
+        // Always try Uint8ClampedArray first, as it should be the most frequent.
+        if (auto* array = dynamicDowncast<JSC::Uint8ClampedArray>(bufferView))
+            return std::forward<F>(f)(*array);
+        if (auto* array = dynamicDowncast<JSC::Float16Array>(bufferView))
+            return std::forward<F>(f)(*array);
+        RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+
+    template <typename F>
+    auto visit(F&& f) const
+    {
+        return visit(m_bufferView.get(), std::forward<F>(f));
+    }
+
+    Ref<JSC::Uint8ClampedArray> asUint8ClampedArray() const;
+    Ref<JSC::Float16Array> asFloat16Array() const;
+
+    JSC::ArrayBufferView& getArrayBufferView() const { return m_bufferView.get(); }
+    JSC::ArrayBufferView* operator->() const { return &getArrayBufferView(); }
+
     auto length() const
     {
-        Ref bufferView = m_bufferView;
-        return bufferView->length();
+        return visit([](const auto& array) { return array.length(); });
     }
 
-    // FIXME: Remove these temporary functions.
-    Uint8ClampedArray& get() const { return m_bufferView.get(); }
-    auto data() const
+    // Needed by `toJS<IDLUnion<IDLUint8ClampedArray, ...>, const ImageDataArray&>()`
+    using Variant = std::variant<RefPtr<JSC::Uint8ClampedArray>, RefPtr<JSC::Float16Array>>;
+    operator Variant() const
```

Comment:
As the comment above says, it's needed by `toJs`. Without it, we get this compilation error:
> no viable conversion from 'const WebCore::ImageDataArray' to 'const ImplementationType' (aka 'const std::variant<RefPtr<JSC::GenericTypedArrayView<JSC::Uint8ClampedAdaptor>>, RefPtr<JSC::GenericTypedArrayView<JSC::Float16Adaptor>>>')

I think it's because in ImageData.idl we have `typedef (Uint8ClampedArray or Float16Array) ImageDataArray;`, this eventually generates JS binding code that expects ImageData.data to be an `IDLUnion` which corresponds to a C++ `std::variant`.

There may be other ways to deal with this, through custom JS converters, but here it seemed easy enough to just provide the `std::variant` that generated JS bindings want.

- Can we keep this for now, and revisit later if you think it would be worth it?
- Should I add more details in the comment above?
- I could potentially hide it in the `private:` section, with the appropriate `friend`ship to let `toJS` use it; this would avoid random uses of this one strange function. What do you think?